File: binchicken/workflow/scripts/evaluate.py
#!/usr/bin/env python3
###################
### evaluate.py ###
###################
# Author: Samuel Aroney

import polars as pl
import os
import argparse
from binchicken.common import parse_snake_dict

SINGLEM_COLUMNS = {
    "gene": str,
    "sample": str,
    "sequence": str,
    "num_hits": int,
    "coverage": float,
    "taxonomy": str,
}

TARGET_COLUMNS = SINGLEM_COLUMNS | {
    "target": int,
}
APPRAISE_COLUMNS = SINGLEM_COLUMNS | {
    "found_in": str,
}

CLUSTER_COLUMNS = {
    "samples": str,
    "length": int,
    "total_targets": float,
    "total_size": float,
    "recover_samples": str,
    "coassembly": str,
}
EDGE_COLUMNS = {
    "style": str,
    "cluster_size": int,
    "samples": str,
    "target_ids": str,
}

OUTPUT_COLUMNS = {
    "coassembly": str,
    "gene": str,
    "sequence": str,
    "genome": str,
    "target": str,
    "found_in": str,
    "source_samples": str,
    "source_num_hits": int,
    "source_coverage": float,
    "taxonomy": str,
    }
SUMMARY_COLUMNS = {
    "coassembly": str,
    "statistic": str,
    "within": str,
    "match": int,
    "nonmatch": int,
    "total": int,
    "match_percent": float,
    }

# Example shell directive for Snakemake:
# shell:
# """
# python3 binchicken/workflow/scripts/evaluate.py \
#   --target-otu-table {params.target_otu_table} \
#   --binned-otu-table {params.binned_otu_table} \
#   --elusive-clusters {params.elusive_clusters} \
#   --elusive-edges {params.elusive_edges} \
#   --recovered-otu-table {input.recovered_otu_table} \
#   --recovered-bins {params.recovered_bins} \
#   --matched-hits {output.matched_hits} \
#   --novel-hits {output.novel_hits} \
#   --summary-stats {output.summary_stats} \
#   --threads {threads} \
#   --log {log}
# """

def evaluate(target_otu_table, binned_otu_table, elusive_clusters, elusive_edges, recovered_otu_table, recovered_bins):

    print(f"Polars using {str(pl.thread_pool_size())} threads")

    if len(recovered_otu_table) == 0:
        empty_output = pl.DataFrame(schema=OUTPUT_COLUMNS)
        return empty_output, empty_output, pl.DataFrame(schema=SUMMARY_COLUMNS)

    # Load elusive clusters and edges (to match targets to coassemblies, with duplicates)
    sample_coassemblies = (
        elusive_clusters
        .select(pl.col("samples").str.split(","), "coassembly")
        .explode("samples")
    )

    elusive_edges = (
        elusive_edges
        .with_columns(
            pl.col("samples").str.split(",").list.eval(
                        pl.when(pl.all_horizontal(pl.element().is_in(sample_coassemblies.get_column("samples"))))
                        .then(pl.element())
                        .otherwise(pl.element().str.replace(r"(_|\.)1$", ""))
                        ),
            pl.col("samples").hash().alias("samples_hash")
            )
        .explode("samples")
        .join(sample_coassemblies, on="samples")
        .group_by("samples_hash", "coassembly")
        .agg(
            pl.first("target_ids"),
            pl.len().alias("count"),
            pl.first("cluster_size"),
            pl.col("samples").unique(),
            )
        .filter(pl.col("count") == pl.col("cluster_size"))
        .with_columns(target = pl.col("target_ids").str.split(","))
        .explode("target")
    )

    # Create otu table with original sequence, samples present, cluster id, target id and associated coassemblies
    sample_edges = (
        elusive_edges
        .group_by("coassembly", "target")
        .agg(
            source_samples = pl.col("samples")
                .flatten()
                .unique()
            )
        .explode("source_samples")
    )

    elusive_otu_table = (
        target_otu_table
        .with_columns(
            pl.col("target").cast(str),
            pl.lit(None).cast(str).alias("found_in"),
            source_samples =
                pl.when(pl.col("sample").is_in(sample_coassemblies.get_column("samples")))
                    .then(pl.col("sample"))
                .otherwise(pl.col("sample").str.replace(r"(_|\.)1$", "")),
            )
        .join(sample_edges, how="inner", on=["target", "source_samples"])
        .group_by("gene", "sequence", "coassembly", "taxonomy", "found_in", "target")
        .agg(
            pl.col("source_samples").sort().str.concat(","),
            source_num_hits = pl.sum("num_hits"),
            source_coverage = pl.sum("coverage"),
            )
    )

    # Add binned otu table to above with target NA
    nontarget_otu_table = (
        pl.concat([
            binned_otu_table
                .with_columns(target = pl.lit(None).cast(str)),
            target_otu_table
                .join(elusive_otu_table, on=["gene", "sequence"], how="anti")
                .drop("target")
                .with_columns(
                    found_in = pl.lit(None).cast(str),
                    target = pl.lit("None"),
                    ),
            ])
        .select([
            pl.col("sample").str.replace(r"_1$", "").str.replace(r"\.1$", ""),
            "gene", "sequence", "taxonomy", "found_in", "num_hits", "coverage", "target"
            ])
        .join(sample_coassemblies, left_on="sample", right_on="samples", how="left", coalesce=True)
        .drop_nulls("coassembly")
        .group_by(["gene", "sequence", "coassembly"])
        .agg([
            pl.first("taxonomy"),
            pl.first("found_in"),
            pl.first("target"),
            pl.col("sample").unique().sort().str.concat(",").alias("source_samples"),
            pl.sum("num_hits").alias("source_num_hits"),
            pl.sum("coverage").alias("source_coverage"),
            ])
        .unique()
    )

    haystack_otu_table = pl.concat([elusive_otu_table, nontarget_otu_table])

    # Load recovered otu table and join elusive and nontarget sequences
    recovered_otu_table = (
        recovered_otu_table
        .with_columns(
            pl.col("sample")
                .str.split("-")
                .list.to_struct(upper_bound=2, fields=["coassembly", "genome"])
            )
        .unnest("sample")
    )

    combined_otu_table = (
        recovered_otu_table
        .join(
            haystack_otu_table, on=["coassembly", "gene", "sequence"], how="full", coalesce=True, suffix="old"
            )
        .select(
            "coassembly", "gene", "sequence", "genome", "target", "found_in",
            "source_samples", "source_num_hits", "source_coverage",
            pl.when(pl.col("taxonomy").is_null())
                .then(pl.col("taxonomyold"))
                .otherwise(pl.col("taxonomy"))
                .alias("taxonomy"),
            )
        .filter(
            (pl.col("coassembly").is_in(pl.lit(recovered_otu_table["coassembly"]))) &
            # Choose sequences where genome is present (from recovered) and/or target is present (from unbinned targets)
            ((pl.col("genome").is_not_null()) | (pl.col("target").is_not_null()))
            )
    )

    matches = (
        combined_otu_table
        .filter(
            ~pl.all_horizontal(pl.col(["target", "found_in", "source_samples"]).is_null())
            )
        .with_columns(
            target = pl.when(pl.col("target") == "None").then(pl.lit(None)).otherwise(pl.col("target"))
            )
    )

    unmatched = (
        combined_otu_table
        .filter(
            pl.all_horizontal(pl.col(["target", "found_in", "source_samples"]).is_null())
            )
    )

    # Summarise recovery stats
    summary = summarise_stats(matches, combined_otu_table, recovered_bins)

    return matches, unmatched, summary

def summarise_stats(matches, combined_otu_table, recovered_bins):
    recovered_hits = (
        combined_otu_table
        .filter(
            pl.col("genome").is_not_null()
            )
        .with_columns(
            target = pl.when(pl.col("target") == "None").then(pl.lit(None)).otherwise(pl.col("target"))
            )
    )

    summary = (
        matches
        .filter(pl.col("target").is_not_null())
        .with_columns(
            pl.when(pl.col("genome").is_not_null())
                .then(pl.lit("match"))
                .otherwise(pl.lit("nonmatch"))
                .alias("status")
            )
        .group_by([
            "coassembly", "status"
            ])
        .agg(
            pl.col("sequence").unique().len().alias("sequences"),
            pl.col("genome").unique().is_not_null().sum().alias("bins"),
            pl.col("taxonomy").unique().is_not_null().sum().alias("taxonomy"),
            )
        .join(
            # Duplicate sequences are counted multiple times to give a proportion at bin level
            recovered_hits
                .with_columns(
                    pl.when(
                        pl.col("found_in").is_not_null()
                    ).then(pl.lit("match")).otherwise(pl.lit("nonmatch")).alias("status")
                    )
                .group_by(["coassembly", "status"])
                .agg(pl.col("sequence").len().alias("nontarget_bin_sequences")),
            on=["coassembly", "status"], how="full", coalesce=True
            )
        .join(
            # Duplicate sequences are counted multiple times to give a proportion at bin level
            recovered_hits
                .with_columns(
                    pl.when(
                        pl.all_horizontal(pl.col(["target", "found_in"]).is_null()) & (pl.col("source_samples").is_not_null())
                    ).then(pl.lit("match")).otherwise(pl.lit("nonmatch")).alias("status")
                    )
                .group_by(["coassembly", "status"])
                .agg(pl.col("sequence").len().alias("nontarget_unbin_sequences")),
            on=["coassembly", "status"], how="full", coalesce=True
            )
        .join(
            # Duplicate sequences are counted multiple times to give a proportion at bin level
            recovered_hits
                .with_columns(
                    pl.when(
                        pl.all_horizontal(pl.col(["target", "found_in", "source_samples"]).is_null())
                    ).then(pl.lit("match")).otherwise(pl.lit("nonmatch")).alias("status")
                    )
                .group_by(["coassembly", "status"])
                .agg(pl.col("sequence").len().alias("novel_sequences")),
            on=["coassembly", "status"], how="full", coalesce=True
            )
        .unpivot(index=["coassembly", "status"], variable_name="statistic", value_name="value")
        .fill_null(0)
    )

    recovered_counts = (
        pl.from_dict(recovered_bins)
        .unpivot(variable_name="name")
        .with_columns(
            pl.col("name").str.split("-").list.to_struct(upper_bound=2, fields=["coassembly", "genome"])
            )
        .unnest("name")
        .group_by("coassembly")
        .agg(pl.len().alias("n"))
        .with_columns(
            pl.lit("match").alias("status"),
            pl.lit("bins").alias("statistic"),
            )
        .join(summary.rename({"value": "match"}), on=["coassembly", "status", "statistic"], how="left", coalesce=True)
        .select(
            "coassembly",
            pl.lit("nonmatch").alias("status"),
            "statistic",
            pl.col("n").alias("value") - pl.col("match")
            )
    )

    summary = (
        pl.concat([
            summary.filter((pl.col("statistic") != "bins") | (pl.col("status") != "nonmatch")),
            recovered_counts
            ])
        .pivot(index=["coassembly", "statistic"], values="value", on="status", aggregate_function=None)
        .select(
            "coassembly", "statistic",
            pl.when(pl.col("statistic").is_in(["sequences", "taxonomy"])).then(pl.lit("targets")).otherwise(pl.lit("recovery")).alias("within"),
            "match", "nonmatch",
            pl.col("match").alias("total") + pl.col("nonmatch"),
            (pl.col("match") / (pl.col("match") + pl.col("nonmatch")) * 100).round(2).alias("match_percent"),
            )
        .sort(["coassembly", "within", "statistic"], descending=[False, True, False])
    )

    return summary

def main():
    parser = argparse.ArgumentParser(description="Evaluate pipeline script.")
    parser.add_argument("--target-otu-table", required=True, help="Path to target OTU table")
    parser.add_argument("--binned-otu-table", required=True, help="Path to binned OTU table")
    parser.add_argument("--elusive-clusters", required=True, help="Path to elusive clusters file")
    parser.add_argument("--elusive-edges", required=True, help="Path to elusive edges file")
    parser.add_argument("--recovered-otu-table", required=True, help="Path to recovered OTU table")
    parser.add_argument("--recovered-bins", required=True, type=parse_snake_dict, help="Recovered bins dictionary (as string)")
    parser.add_argument("--matched-hits", required=True, help="Path to output matched hits file")
    parser.add_argument("--novel-hits", required=True, help="Path to output novel hits file")
    parser.add_argument("--summary-stats", required=True, help="Path to output summary stats file")
    parser.add_argument("--threads", type=int, default=1, help="Number of threads for Polars")
    parser.add_argument("--log", default=None, help="Log file path")
    args = parser.parse_args()

    os.environ["POLARS_MAX_THREADS"] = str(args.threads)
    import polars as pl

    if args.log:
        import logging
        logging.basicConfig(
            filename=args.log,
            level=logging.INFO,
            format='%(asctime)s %(levelname)s: %(message)s',
            datefmt='%Y/%m/%d %I:%M:%S %p'
        )
    else:
        import logging
        logging.basicConfig(
            level=logging.INFO,
            format='%(asctime)s %(levelname)s: %(message)s',
            datefmt='%Y/%m/%d %I:%M:%S %p'
        )

    target_otu_table = pl.read_csv(args.target_otu_table, separator="\t", schema_overrides=TARGET_COLUMNS)
    binned_otu_table = pl.read_csv(args.binned_otu_table, separator="\t", schema_overrides=APPRAISE_COLUMNS)
    elusive_clusters = pl.read_csv(args.elusive_clusters, separator="\t", schema_overrides=CLUSTER_COLUMNS)
    elusive_edges = pl.read_csv(args.elusive_edges, separator="\t", schema_overrides=EDGE_COLUMNS)
    recovered_otu_table = pl.read_csv(args.recovered_otu_table, separator="\t", schema_overrides=SINGLEM_COLUMNS)

    matches, unmatched, summary = evaluate(target_otu_table, binned_otu_table, elusive_clusters, elusive_edges, recovered_otu_table, args.recovered_bins)
    matches.write_csv(args.matched_hits, separator="\t")
    unmatched.write_csv(args.novel_hits, separator="\t")
    summary.write_csv(args.summary_stats, separator="\t")

if __name__ == "__main__":
    main()

File: binchicken/workflow/scripts/evaluate.R
#!/usr/bin/env Rscript
##################
### evaluate.R ###
##################
# Author: Samuel Aroney

# Example shell directive for Snakemake:
# shell:
# """
# Rscript binchicken/workflow/scripts/evaluate.R \
#   --matched-hits {input.matched_hits} \
#   --novel-hits {input.novel_hits} \
#   --cluster-summary {input.cluster_summary} \
#   --summary-stats {input.summary_stats} \
#   --coassemble-summary {params.coassemble_summary} \
#   --plots-dir {output.plots_dir} \
#   --summary-table {output.summary_table} \
#   --test {config[test]}
# """

suppressPackageStartupMessages({
  library(optparse)
  library(gt)
  library(cowplot)
  library(tidyverse)
})

option_list <- list(
  make_option("--matched-hits", type="character", help="Path to matched hits file"),
  make_option("--novel-hits", type="character", help="Path to novel hits file"),
  make_option("--cluster-summary", type="character", default=NULL, help="Path to cluster summary file (optional)"),
  make_option("--summary-stats", type="character", help="Path to summary stats file"),
  make_option("--coassemble-summary", type="character", help="Path to coassemble summary file"),
  make_option("--plots-dir", type="character", help="Path to output plots directory"),
  make_option("--summary-table", type="character", help="Path to output summary table file"),
  make_option("--test", type="logical", default=FALSE, help="Test mode (True/False)")
)

opt <- parse_args(OptionParser(option_list=option_list))

test <- isTRUE(opt$test)
if (test) sink(file("/dev/null", "w"), type = "message")

Sys.setenv(OPENSSL_CONF = "/dev/null")
if (!webshot::is_phantomjs_installed()) webshot::install_phantomjs()

# Inputs from argparse
matched_hits_path <- opt$`matched-hits`
novel_hits_path <- opt$`novel-hits`
cluster_summary_path <- opt$`cluster-summary`
summary_stats_path <- opt$`summary-stats`
coassemble_summary_path <- opt$`coassemble-summary`
main_dir <- opt$`plots-dir`
summary_table_path <- opt$`summary-table`

# Load inputs
matched_hits <- read_tsv(matched_hits_path)
novel_hits <- read_tsv(novel_hits_path)
coassemble_summary <- read_tsv(coassemble_summary_path)
summary_stats <- read_tsv(summary_stats_path)

if (is.null(cluster_summary_path) || cluster_summary_path == "" || cluster_summary_path == "NULL") {
    cluster_summary <- tibble(type = c("original", coassemble_summary$coassembly), clusters = NA_real_)
} else {
    cluster_summary <- read_csv(cluster_summary_path, col_names = c("type", "clusters"))
}

#################
### Functions ###
#################
plot_bars <- function(df, output_dir) {
    dir.create(output_dir, recursive = TRUE)

    bar_layers <- list(
        geom_col(),
        scale_fill_brewer(breaks = c("recovered", "missed"), palette = "Set2"),
        ylab("SingleM sequences"),
        coord_flip(),
        theme_cowplot(),
        theme(axis.title.y = element_blank())
    )

    df %>%
        group_by(Phylum) %>%
        summarise(
            recovered = sum(!is.na(genome)),
            missed = sum(is.na(genome))
        ) %>%
        pivot_longer(-Phylum, names_to = "outcome", values_to = "count") %>%
        ggplot(aes(reorder(Phylum, desc(count), FUN = sum), count, fill = outcome)) +
        bar_layers
    ggsave(str_c(output_dir, "/phylum_recovered.png"), dpi = 900, height = 8, width = 13)

    df %>%
        group_by(gene) %>%
        summarise(
            recovered = sum(!is.na(genome)),
            missed = sum(is.na(genome))
        ) %>%
        pivot_longer(-gene, names_to = "outcome", values_to = "count") %>%
        mutate(gene_number = map_int(gene, ~ str_split(., "\\.")[[1]][2] %>% as.integer())) %>%
        ggplot(aes(reorder(gene, gene_number, FUN = sum), count, fill = outcome)) +
        bar_layers
    ggsave(str_c(output_dir, "/gene_recovered.png"), dpi = 900, height = 8, width = 13)

    return(invisible(NULL))
}

plot_coassembly <- function(coassembly) {
    analysis %>%
        filter(coassembly == {{coassembly}}) %>%
        plot_bars(str_c(main_dir, coassembly, sep = "/"))

    return(invisible(NULL))
}

################
### Plotting ###
################
dir.create(main_dir, recursive = TRUE)
taxonomy_groups <- c("Root", "Domain", "Phylum", "Class", "Order", "Family", "Genus", "Species")

analysis <- matched_hits %>%
    filter(!is.na(target)) %>%
    separate(taxonomy, into = taxonomy_groups, sep = "; ", remove = FALSE)

if (test) warning(analysis)

plot_bars(analysis, str_c(main_dir, "/combined"))

coassembly_plots <- analysis %>%
    distinct(coassembly) %>%
    pull(coassembly) %>%
    map(plot_coassembly)

####################
### Table output ###
####################
target_summary <- summary_stats %>%
    filter(statistic != "taxonomy") %>%
    pivot_wider(id_cols = coassembly, names_from = statistic, values_from = match)

target_totals <- summary_stats %>%
    filter(!statistic %in% c("nontarget_unbin_sequences", "novel_sequences", "taxonomy")) %>%
    pivot_wider(id_cols = coassembly, names_from = statistic, values_from = total) %>%
    rename(total_targets = sequences, total_bins = bins, total_recovered = nontarget_bin_sequences)

target_percentage <- summary_stats %>%
    filter(!statistic %in% c("bins", "nontarget_bin_sequences", "nontarget_unbin_sequences", "novel_sequences", "taxonomy")) %>%
    pivot_wider(id_cols = coassembly, names_from = statistic, values_from = match_percent) %>%
    rename(perc_targets = sequences)

original_clusters <- cluster_summary %>%
    filter(type == "original") %>%
    pull(clusters)

cluster_info <- cluster_summary %>%
    pivot_wider(names_from = type, values_from = clusters) %>%
    pivot_longer(-original, names_to = "coassembly", values_to = "clusters") %>%
    mutate(novel_clusters = clusters - original) %>%
    select(coassembly, novel_clusters)

if (!"unmapped_size" %in% colnames(coassemble_summary)) coassemble_summary$unmapped_size <- NA_real_

summary_table <- coassemble_summary %>%
    filter(coassembly %in% analysis$coassembly) %>%
    select(-c(samples, total_targets)) %>%
    left_join(target_summary) %>%
    left_join(target_totals) %>%
    left_join(target_percentage) %>%
    left_join(cluster_info) %>%
    mutate(
        sequences = pmap_chr(list(sequences, total_targets, perc_targets), ~ str_c(..1, "/", ..2, " (", ..3, "%)")),
        bins = map2_chr(bins, total_bins, ~ str_c(.x, " (", .y, " total)")),
        total_size = total_size / 10**9,
        unmapped_size = unmapped_size / 10**9,
        ) %>%
    select(coassembly, length, total_size, unmapped_size, bins, sequences, nontarget_bin_sequences, nontarget_unbin_sequences, novel_sequences, novel_clusters) %>%
    gt() %>%
    tab_spanner(
        label = "Gbp",
        columns = c(total_size, unmapped_size)
    ) %>%
    tab_spanner(
        label = "Recovered sequences",
        columns = c(sequences, nontarget_bin_sequences, nontarget_unbin_sequences, novel_sequences)
    ) %>%
    fmt_integer(c(length, total_size, unmapped_size, nontarget_bin_sequences, nontarget_unbin_sequences, novel_sequences, novel_clusters)) %>%
    cols_label(
        coassembly = "coassembly",
        length = "samples",
        total_size = "size",
        unmapped_size = "update",
        bins = "target bins",
        sequences = "targets",
        nontarget_unbin_sequences = "non-targets",
        nontarget_bin_sequences = "prior binned",
        novel_sequences = "novel",
        novel_clusters = "novel clusters"
    ) %>%
    tab_header(
        title = "Bin Chicken coassembly evaluation",
        subtitle = str_c("Original genomes had ", original_clusters, " clusters")
        ) %>%
    tab_style(
        style = cell_fill(),
        locations = cells_body(columns = c(length, unmapped_size, sequences, novel_sequences))
        )

if (is.null(cluster_summary_path)) summary_table <- summary_table %>% cols_hide(novel_clusters)

gtsave(summary_table, summary_table_path)

# Save R image for further processing
save.image(file = str_c(main_dir, "/evaluate.RData"))
